### PR TITLE
Redesign user auth

### PR DESCRIPTION
**Pull request by nirinchev**
## What, How & Why?

This aligns the authentication API with the other SDKs.

## ☑️ ToDos

* [x] 📝 Changelog entry
* [x] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [ ] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: lib/user-methods.js [hunk 1]
```diff
@@ -609,10 +515,64 @@ const instanceMethods = {
...
+        return new Credentials('azuread', token)
+    },
+
+    jwt(token) {
+        return new Credentials('jwt', token);
+    },
+
+    adminToken(token) {
+        return new Credentials('adminToken', token);
+    },
```
Comment (kraenhansen) [inline lib/user-methods.js]: I guess we could use `checkTypes` for `nickname`, `azureAD`, `jwt` and `adminToken` as well?

File: lib/user-methods.js [hunk 2]
```diff
@@ -609,10 +515,64 @@ const instanceMethods = {
...
+    nickname(value, isAdmin) {
+        return new Credentials('nickname', value, { is_admin: isAdmin || false });
+    },
+
+    azureAD(token) {
+        return new Credentials('azuread', token)
+    },
+
+    jwt(token) {
+        return new Credentials('jwt', token);
```
Comment (kraenhansen) [inline lib/user-methods.js]: Should this take an optional `provider` too?

File: docs/sync.js [hunk 3]
```diff
@@ -288,58 +288,105 @@ class IncompatibleSyncedRealmError {
...
+    /**
+     * Gets additional user information associated with the credentials.
+     * @returns {object} A dictionary, containing the additional information.
+     */
+    get userInfo();
+
+}
+
+/**
+ * Class for logging in and managing Sync users.
```
Comment (kneth) [inline docs/sync.js]: Just leave out "logging in and".

**Comment by kneth** [comment]
@nirinchev I have connected your PR with #1579 which I believe is the correct issue. To me, it is important that we get something which is as close to our other SDKs as possible.

**Comment by nirinchev** [comment]
I agree - this very closely mirrors the .NET API, which if I remember correctly is aligned with Cocoa and Java (maybe with a few casing differences). @cmelchior can you do a high level review and see if the new API diverges in significant ways from Java?

File: lib/index.d.ts [hunk 4]
```diff
@@ -290,6 +290,21 @@ declare namespace Realm.Sync {
         isAdmin: boolean;
     }
 
+    class Credentials {
```
Comment (cmelchior) [inline lib/index.d.ts]: In Java this is called `SyncCredentials` and `SyncUser`, here it is called `Credentials` and `User` but I guess that is due it being inside the Realm.Sync namespace.

File: lib/index.d.ts [hunk 5]
```diff
@@ -290,6 +290,21 @@ declare namespace Realm.Sync {
...
+        static usernamePassword(username: string, password: string, createUser?: boolean): Credentials;
+        static facebook(token: string): Credentials;
+        static google(token: string): Credentials;
+        static anonymous(): Credentials;
+        static nickname(value: string, isAdmin?: boolean): Credentials;
+        static azureAD(token: string): Credentials;
+        static jwt(token: string): Credentials;
+        static adminToken(token: string): Credentials;
+
+        readonly provider: string;
```
Comment (cmelchior) [inline lib/index.d.ts]: Java uses the term `identityProvider`

File: lib/index.d.ts [hunk 6]
```diff
@@ -290,6 +290,21 @@ declare namespace Realm.Sync {
...
+        readonly token: string;
```
Comment (cmelchior) [inline lib/index.d.ts]: And this is called `userIdentifier`

File: lib/index.d.ts (hunk 5 above)
Comment (nirinchev) [reply lib/index.d.ts]: .NET uses `IdentityProvider` as well, Cocoa uses `provider`, but Cocoa's version is internal, so people can't actually access it. I guess, I'll change it to align with Java and .NET

File: lib/index.d.ts (hunk 6 above)
Comment (nirinchev) [reply lib/index.d.ts]: In .NET and Cocoa we use `token`, although as mentioned, Cocoa's properties are private. I'll speak with Kenneth and see which one he likes better - I don't have a strong opinion on the matter :)

**Comment by bmunkholm** [comment]
@nirinchev The "pipeline" in zenhub tracks the state - doing that in the subject with "[WIP]" is redundant and not a scheme that is used by others either, so I've removed "[WIP]" :-)

File: CHANGELOG.md [hunk 7]
```diff
@@ -7,7 +7,23 @@ X.Y.Z Release notes
 * Realm Object Server: 3.0.0 or later
 
 ### Breaking changes
-* None.
+* The authentication API have been completely revamped.
```
Comment (kneth) [inline CHANGELOG.md]: "have" -> "has"

File: CHANGELOG.md [hunk 8]
```diff
@@ -7,7 +7,23 @@ X.Y.Z Release notes
...
+  * The following methods have been removed:
```
Comment (kneth) [inline CHANGELOG.md]: "removed" -> "deprecated and will be removed at next major version"

File: docs/sync.js [hunk 9]
```diff
@@ -288,58 +288,120 @@ class IncompatibleSyncedRealmError {
...
+    /**
+     * Creates credentials based on an Active Directory login.
+     * @param {string} token An access token, obtained by logging into Azure Active Directory.
+     * @return {Credentials} An instance of `Credentials` that can be used in {@linkcode Realm.Sync.User.login|User.login}.
+     */
+    static azureAD(token) {};
+
+    /**
+     * Creates credentials based on a JWT login.
+     * @param {string} token A Json Web Token, that will be validated against the server's configured rules.
```
Comment (kneth) [inline docs/sync.js]: "Json" -> "JSON"

File: CHANGELOG.md (hunk 7 above)
Comment (nirinchev) [reply CHANGELOG.md]: I was using the plural version of API 🤓

**Comment by kneth** [comment]
The CI errors are `The @deprecated, tag is not a known tag. File: sync.js` - which is strange as it is a [valid tag](http://usejsdoc.org/tags-deprecated.html).

**Pull request merged**